azure-iot-device/azure/iot/device/common/pipeline/pipeline_stages_base.py

```diff
@@ -969,223 +930,336 @@ def do_retry():
-            else:
-                # TODO: we are missing test cases for this condition
-                logger.info(
-                    "{}({}): State changes {}->LOGICALLY_CONNECTED.  Adding to wait list and sending new connect op down".format(
+                    "{}({}): State is {} - waiting for in-progress operation to finish".format(
                         self.name, op.name, self.state
                     )
                 )
-                self.state = ReconnectState.LOGICALLY_CONNECTED
-                # We don't send this op down.  Instead, we send a new connect op down.  This way,
-                # we can distinguish between connect ops that we're handling (they go into the
-                # queue) and connect ops that we are sending down.
-                #
-                # Once we finally connect, we only have to complete the ops in the queue and we
-                # never have to worry about completing the op that we sent down.  The code is much
-                # cleaner this way, especially when you take retries into account, trust me.
-                self.waiting_connect_ops.append(op)
-                self._send_new_connect_op_down()
-
-        elif isinstance(op, pipeline_ops_base.DisconnectOperation):
-            if self.state == ReconnectState.WAITING_TO_RECONNECT:
-                logger.info(
-                    "{}({}): State changes {}->LOGICALLY_DISCONNECTED.  Canceling waiting ops and sending disconnect down.".format(
-                        self.name, op.name, self.state
-                    )
-                )
-                self.state = ReconnectState.LOGICALLY_DISCONNECTED
-                self._clear_reconnect_timer()
-                self._complete_waiting_connect_ops(
-                    pipeline_exceptions.OperationCancelled("Explicit disconnect invoked")
-                )
-                # TODO: There is a bug here. If in a WAITING_TO_RECONNECT state, and an explicit
-                # Disconnect occurs, it is completed here rather than going all the way down. This
-                # means that any in-flight operations are NOT cancelled.
-                op.complete()
+                self.waiting_ops.put_nowait(op)
+                self.state_lock.release()
 
             else:
-                logger.info(
-                    "{}({}): State changes {}->LOGICALLY_DISCONNECTED.  Sending op down.".format(
-                        self.name, op.name, self.state
-                    )
-                )
-                self.state = ReconnectState.LOGICALLY_DISCONNECTED
-                self.send_op_down(op)
+                if isinstance(op, pipeline_ops_base.ConnectOperation):
+                    if self.state is ReconnectState.CONNECTED:
+                        logger.debug(
+                            "{}({}): State is already CONNECTED. Sending op down".format(
+                                self.name, op.name
+                            )
```

Comment:
This is an interesting choice.  Sending the ConnectOperation down when self.state says that you're already connect impies that you don't trust self.state to be correct.  I can see why you migth not, but if you're passing down a ConnectOperation, wouldn't you go into ReconnectState.CONNECTING?

Reply:
This is one of the trickier points - my overall philosophy here is that I want the operations to go down the pipeline, although the ConnectionLockStage that comes next gets in the way of that, and I didn't want to touch that in this PR. The ConnectionLockStage is going to immediately complete the operation successfully without letting it ever hit the transport, and in such case, it would cause the state to get stuck forever in CONNECTING (because it is only moved to a stable state by a ConnectedEvent, which would never happen). Conversely, if the ConnectionLockStage fails it, because of how aggressive it can be with it's cancellation policy, we would end up in a potentially erroneous DISCONNECTED state as a result of the callback that was applied to the ConnectOperation.

For now this seemed the safest thing to do - keep the state where we think it probably should be, and let Events fired from the transport sort it out. However, when revisiting the ConnectionLockStage, this would need to be revisited as well.

I have added a comment to reflect this.



The issue with putting it into a CONNECTING state is that if that operation were to fail, it would (currently) then change the state, quite possibly erroneously, to DISCONNECTED. Certainly, if the ConnectionLockStage gets altered t